android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt

```diff
@@ -0,0 +1,167 @@
+ *
+ * @param <C> the type of Configuration this Interactor can expect to push to its [Router].
+ * @param <V> the type of [RibView].
+ **/
+abstract class Interactor<C : Parcelable, V : RibView>(
+    protected val router: Router<C, V>,
+    private val disposables: Disposable?
+) : LifecycleScopeProvider<InteractorEvent>, LifecycleOwner {
+
+    private val ribLifecycleRegistry = LifecycleRegistry(this)
+    private val viewLifecycleRegistry = LifecycleRegistry(this)
+
+    // todo these are leftovers, reconsider them
+    private val behaviorRelay = BehaviorRelay.create<InteractorEvent>()
+    private val lifecycleRelay = behaviorRelay.toSerialized()
+
+    val isAttached: Boolean
+        get() = behaviorRelay.value == ACTIVE
+
+    // todo these are leftovers, reconsider them
+    override fun lifecycle(): Observable<InteractorEvent> =
+        lifecycleRelay.hide()
+
+    /**
+     * Called when attached. The presenter will automatically be added when this happens.
+     *
+     * @param savedInstanceState the saved [Bundle].
+     */
+    @CallSuper
+    protected fun didBecomeActive(savedInstanceState: Bundle?) {
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
+        didBecomeActive(ribLifecycleRegistry, savedInstanceState)
+    }
+
+    protected open fun didBecomeActive(ribLifecycle: Lifecycle, savedInstanceState: Bundle?) {
+    }
+
+    @CallSuper
+    fun onViewCreated(view: V) {
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
+        onViewCreated(view, viewLifecycleRegistry)
+    }
+
+    protected open fun onViewCreated(view: V, viewLifecycle: Lifecycle) {
+    }
+
+    fun onViewDestroyed() {
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+    }
+
+    fun onStart() {
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
+    }
+
+    // todo call this when removed from view
+    fun onStop() {
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
+    }
+
+    fun onResume() {
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
+    }
+
+    // todo call this when removed from view
+    fun onPause() {
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
+    }
+
+    /**
+     * Called when detached. The [Interactor] should do its cleanup here.
+     */
+    fun willResignActive() {
+        viewLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY) // todo probably this is not needed?
+        ribLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+        disposables?.dispose()
```

Comment:
You can use binder with view lifecycle for this. It will prevent feature from receiving any signals when lifecycle has ended.

Reply:
Good point, will work.
Looks like we should use viewLifecycle almost all the time?
Because we usually don't want to catch events when our screen is in background.

Reply:
Depends, sometimes you want to have start/stop there

Reply:
@Nublo This came up early on and the group decision was that there's a need to keep stuff alive and listening while it's attached as a logical child (still in the backstack) but detached from view. In case this is not needed, you can still bind to view lifecycle as @ShikaSD said. Also it's not in MVP, but a simple flag can be added to rib factories in routing actions like "keepAlive", where if you set it false the whole thing will actually be killed, not just its view. Since ShrinkToBundles Wish already does that and replaces ribs in the back stack with their corresponding Bundles, it should be trivial to implement if we need it.